### ISSUE TITLE
What is this software about?

### ISSUE THREAD

**Issue by dreirund** [stsp/thunk_gen#1]
Ahoj,

can you add a `README.md` which describes what this software is about, how to build it and how to use it?

Regards!

**Comment by stsp** [comment]
NP, added.

**Comment by dreirund** [comment]
> *NP, added.*

Thanks.

Now [there is an Arch Linux package for this](https://aur.archlinux.org/packages/thunk_gen-git).

Regards!

**Comment by stsp** [comment]
Thanks!
So I guess they used older fdpp
that includes thunk_gen within?

I would suggest to move the comcom32
dependency from fdpp-git to dosemu2-git
btw.

**Comment by stsp** [comment]
Also I would suggest to update
nasm-segelf, as it no longer
`Conflicts: 	nasm, nasm-elf16, nasm-segelf`
but can rather be installed in parallel
with nasm.

**Comment by severach** [comment]
Moving the comcom32 dep to dosemu2 doesn't make sense. dosemu doesn't need it. fdpp does. Without fdpp, dosemu uses FreeDOS.

**Comment by stsp** [comment]
Crap, those are entirely unrelated.
You can use dosemu2+freedos+comcom64
or dosemu2+fdpp+freecom or any
other combination of such.
The MS-DOS tradition of releasing the
command.com that checks the DOS
version and works only with his "native"
DOS is not here.

Also you are not supposed to second-guess
our defaults. It was only the problem
with arch, when people came here
reporting the freedos-related problems,
and I had to say "unsupported".

**Comment by dreirund** [comment]
So what then?

* `comcom32` as optional dependency for `dosemu2`,
* `fdpp` as optional dependency for `dosemu2`,
* `comcom32` as required dependency for `fdpp`?

**Comment by stsp** [comment]
> comcom32 as optional dependency for dosemu2,
> comcom64 as optional dependency for dosemu2,

It would make sense to have one
of them a mandatory dependency,
as you are not supposed to run
dosemu2 w/o command.com at all.
There should be at least something.

> fdpp as optional dependency for dosemu2,

I'd say fdpp should be a mandatory.
User can still boot any DOS if he have
one, but there must be something
pre-installed.

> comcom32 as required dependency for fdpp?

No, those are completely unrelated.
fdpp needs nothing, its just a DOS.
It can boot whatever command.com
you have, or no command.com at all
if you specified something else via
`SHELL=`.

**Comment by stsp** [comment]
Its dosemu2 who looks up what DOS
and command.com is installed, provides
the needed config.sys and so on.
So only dosemu2 should have any deps.
Other projects are completely unaware
about an environment and of each other.

**Comment by dreirund** [comment]
On Mon, 25 Mar 2024 11:21:17 -0700, Stas Sergeev
***@***.***> wrote:

> > comcom32 as optional dependency for dosemu2,
> > comcom64 as optional dependency for dosemu2,  
> 
> It would make sense to have one
> of them a mandatory dependency,
> as you are not supposed to run
> dosemu2 w/o command.com at all.
> There should be at least something.
> 
> > fdpp as optional dependency for dosemu2,  
> 
> I'd say fdpp should be a mandatory.
> User can still boot any DOS if he have
> one, but there must be something
> pre-installed.

Having all options mandatory is the wrong way to go.

There can be two options:

* Have all the packages which provide a `command.com` suitable for
  `dosemu2` have added something like `FREEDOS` or `DOSEMUINTERPRETER`
  or `COMMAND.COM` to the `provides` array  
  and add only that as mandatory dependency,
* or really have them all as optional dependency and write a notice in
  `post_install()` in the `$install`-file telling the user that some
  DOS needs to be available. (Especially when also commercial MS DOS
  can do it, I would prefer this, so people are not forced to install
  some thing they do not want.)  
  This would be like "I provide the programme" (`dosemu2` in this
  case), but the user needs some data to work with (like the DOS).


> > comcom32 as required dependency for fdpp?  
> 
> No, those are completely unrelated.

Then `comcom32` can safaly be removed as dependencies from `fdpp`
(-> @severach) and `fdpp-git` (@ me).

**Comment by stsp** [comment]
> Have all the packages which provide a `command.com` suitable for

In debian I can even write
`dep1 | dep2` getting either
one a mandatory (but not both).
If in arch you can do that only
with the virtual providers, then
yes, you can make multiple
packages providing "command.com".
Not sure if you need both
comcom32 and comcom64 being
in such packages, as these are built
from the same source tree, so they
are pretty much equivalent. But if
you package eg comcom64 and
freecom both providing "command.com",
then the user will have the choice
you want him to have.

fdpp is different however, as not
installing it makes a significant
feature set reduction, and, unless
you know what exactly freedos version
to provide - also a completely unsupported
(and broken in most cases) setup.

**Comment by stsp** [comment]
Also installing fdpp doesn't mean
the user can't install freedos or ms-dos.
Installing fdpp makes no obligations,
but gives you the working setup out
of the box.

**Comment by stsp** [comment]
Also note that dosemu2 looks into
drive C for command.com _before_
looking for comcom64 and comcom32.
So installing comcom also makes no
obligations. dosemu2 would boot
whatever it finds on C first, so comcom
just gives an out-of-the-box setup.
Which is why I suggest to make both
fdpp and either of comcoms a mandatory,
and also add an optional deps for
whatever override you want.

**Comment by stsp** [comment]
And also note that we have
scripts for installing various
DOSes, namely:
https://github.com/dosemu2/install-freedos
https://github.com/dosemu2/install-otherdos

They allow user to install whatever
he wants, rather than the packager
doing what you do.
Why to care?
Letting the user to do things on his
own, reduces the problems for the
packager and leaves the software in
a supported state.

**Comment by dreirund** [comment]
On Mon, 25 Mar 2024 19:38:56 -0700, Stas Sergeev
***@***.***> wrote:

> They allow user to install whatever
> he wants, rather than the packager
> doing what you do.
> Why to care?
> Letting the user to do things on his
> own, reduces the problems for the
> packager and leaves the software in
> a supported state.

This for me is an argument to _not_ have _any_ dos or interpreter
listed as required dependency!

**Comment by stsp** [comment]
Install-freedos script doesn't replace
an interpreter. In fact, it needs one to
work, as it provides a bat file.
Install-otherdos installs the DOS in
parallel to an existing setup, so that
you can do
`dosemu -dos <dos_name>`
Which means having a mandatory
dep to an interpreter is orthogonal
to that process and is needed to work
out-of-the-box.

**Comment by stsp** [comment]
I.e. dosemu2 has everything to
install multiple DOSes and choose
one on start.
Why you guys always want to reduce
the functionality just to do the things
on your own, is a bit unclear.

**Comment by severach** [comment]
```
_opt_fdpp=1
# 0 = use freedos
# 1 = use fdpp and comcom32 (boot only, freedos is still used for utilities)
# 2 = use fdpp and freecom
```

dosemu2 does not require comcom32. `_opt_fdpp=0` gets you a ready to run dosemu with a single package build. fdpp, comcom32, and comcom64 need 3-5 more packages. The djgpp packages for comcom32 were large requiring hours to build. The dj64 package group for comcom64 is smaller. The new DOS boot files also change periodically, which is good. Keep building new stuff. With change comes breakage and new package requirements. It's good to have FreeDOS to fall back on to help pinpoint bugs. I set the default to fdpp so your new DOS boot tools get plenty of testing, otherwise everyone would just use FreeDOS.

The reason to limit options is [XKCD: Standards](https://xkcd.com/927/). With dosemu2-git supporting only 3 configurations, we can expect that they are running one of the 3 and I can run it to reproduce any problems. Packaging does not block anyone from run any desired DOS boot and config auto files. If their setup doesn't work, noone else can reproduce it and they should test one of the 3 supported configurations.

The only drawback to a fixed set of configurations is that few will venture outside of those and test dosemu2 more advanced functionality. The con is that they won't even use dosemu2 if it requires 3 days to follow a 50 step process to install. Debian has too many 50 step install guides, which on Arch, the first 45 steps are done and the last 4 are optional.

>Letting the user to do things on his own, reduces the problems for the packager and leaves the software in a supported state.

Letting the user do thing on his own, doesn't change the problems for the packager and leaves the software in a **non** supported state until the testing can be done to get it to a supported state.

**Comment by stsp** [comment]
> The djgpp packages for comcom32 were large requiring hours to build.
> The dj64 package group for comcom64 is smaller.

In this case why none of the 3
options above, have comcom64?
Fixed configurations look better
than fuzzy deps, I agree. This way
you can't get a non-functional
configuration by not installing any
DOS at all or command.com.

**Comment by severach** [comment]
The change from comcom32 to comcom64 isn't complete. I have it working. @dreirund doesn't. When it's all working we'll push the completed packages, otherwise we'll get a flurry of bug reports.

**Comment by stsp** [comment]
Doesn't he have a `noexec` option
on /dev/shm mount?

**Comment by dreirund** [comment]
> Doesn't he have a `noexec` option
> on /dev/shm mount?

I have.

Was not a problem so far.

I also do not use dosemu, I just happen to have it install for "in case I want to use some old DOS software I still have", and in the wake of doing so I maintain two packages.

With `comcom64`, DOSEMU2 failed with `cannot dlopen /dev/shm/dosemu_COMMAND.COM: /dev/shm/dosemu_COMMAND.COM: failed to map segment from shared object`. It did not happen with `comcom32`. At this point I gave up digging further because `comcom32` was working.

Should `/dev/shm` be mounted with executable rights for `comcom64`, but not needed for `comcom32`? That I find difficult to explain to users. And I even don't know by now where it is configured in my ([Artix](http://artixlinux.org/), OpenRC, but not systemd) system (it is _not_ in `/etc/fstab`), and actually overriding such stuff I think is where I would stop for now.

So I get another round of downloading and building `comcom64-git` now just to test this and report back ...

**Comment by severach** [comment]
```
# Arch
$ mount | grep shm
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)

# Manjaro
$ mount | grep shm
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
```

It's all exec here. I didn't do anything to alter these settings.

**Comment by stsp** [comment]
I added the work-around.
Should work now, until some idiot
sets `noexec` also on `/tmp`. Then
it'll break forever.

Note: you need to update the whole
stack to get the work-around. I.e.
dj64dev, comcom64 and dosemu2.

**Comment by stsp** [comment]
> _opt_fdpp=1
> # 0 = use freedos
> # 1 = use fdpp and comcom32 (boot only, freedos is still used for utilities)
> # 2 = use fdpp and freecom

You can add _opt_comcom:
0 = use freecom
1 = use comcom32
2 = use comcom64

_opt_fdpp will then only have 2
values, just as it supposed to be.

Also you can add install-freedos
and install-otherdos scripts to allow
the user to add more DOSes and
select them at startup.
